File: day2/part2/solution.go

package main

import (
	"bufio"
	"fmt"
	"math/big"
	"os"
	"strconv"
	"strings"
)

// Define a struct to hold the range properly
type Range struct {
	Start *big.Int
	End   *big.Int
}

func parseRanges(inputStr string) ([]Range, *big.Int) {
	var ranges []Range
	maxID := big.NewInt(0)
	rangeStrs := strings.Split(strings.TrimSpace(inputStr), ",")
	for _, rStr := range rangeStrs {
		if rStr == "" {
			continue
		}
		parts := strings.Split(rStr, "-")
		start, _ := new(big.Int).SetString(parts[0], 10)
		end, _ := new(big.Int).SetString(parts[1], 10)

		ranges = append(ranges, Range{Start: start, End: end})

		if end.Cmp(maxID) > 0 {
			maxID.Set(end)
		}
	}
	return ranges, maxID
}

func generateInvalidIDs(maxID *big.Int) []*big.Int {
	var invalidIDs []*big.Int
	baseNum := int64(1)

	for {
		baseS := strconv.FormatInt(baseNum, 10)
		
		// Optimization: if the smallest possible repetition (baseS + baseS) is too big, stop.
		firstRepetitionVal, _ := new(big.Int).SetString(baseS + baseS, 10)
		if firstRepetitionVal.Cmp(maxID) > 0 {
			break
		}
		
		currentRepeatedS := baseS
		
		for {
			currentRepeatedS += baseS // Append baseS again
			nInvalid, _ := new(big.Int).SetString(currentRepeatedS, 10)

			if nInvalid.Cmp(maxID) > 0 {
				break // This baseS repetition is too long
			}
			
			invalidIDs = append(invalidIDs, nInvalid)
		}
		
		baseNum++
	}
	return invalidIDs
}

func solve(inputStr string) *big.Int {
	ranges, maxID := parseRanges(inputStr)
	potentialIDs := generateInvalidIDs(maxID)

	foundInvalidIDs := make(map[string]*big.Int) // Using a map as a Set

	for _, invalidID := range potentialIDs {
		for _, r := range ranges {
			if r.Start.Cmp(invalidID) <= 0 && invalidID.Cmp(r.End) <= 0 {
				foundInvalidIDs[invalidID.String()] = invalidID
				break // Found in a range, break from inner loop
			}
		}
	}

	totalSum := big.NewInt(0)
	for _, v := range foundInvalidIDs {
		totalSum.Add(totalSum, v)
	}

	return totalSum
}

func main() {
	inputFilePath := "day2/part2/input.txt" // Relative path to input.txt in the same directory
	file, err := os.Open(inputFilePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file '%s': %v\n", inputFilePath, err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	inputStr := scanner.Text()

	result := solve(inputStr)
	fmt.Printf("The sum of all invalid IDs is: %s\n", result.String())
}

File: day2/part2/solution.rb

require 'set'

def parse_ranges(input_str)
  ranges = []
  max_id = 0
  input_str.strip.split(',').each do |r_str|
    next if r_str.empty?
    start_str, end_str = r_str.split('-')
    start_val = start_str.to_i
    end_val = end_str.to_i
    ranges.push([start_val, end_val])
    max_id = end_val if end_val > max_id
  end
  [ranges, max_id]
end

def generate_invalid_ids(max_id)
  invalid_ids = []
  base_num = 1

  loop do
    base_s = base_num.to_s
    
    # Optimization: if the smallest possible repetition (base_s + base_s) is too big, stop.
    first_repetition_val = (base_s + base_s).to_i
    break if first_repetition_val > max_id
    
    current_repeated_s = base_s
    
    loop do
      current_repeated_s += base_s # Append base_s again
      n_invalid = current_repeated_s.to_i

      break if n_invalid > max_id
      
      invalid_ids.push(n_invalid)
    end
    
    base_num += 1
  end
  invalid_ids
end

def solve(input_str)
  ranges, max_id = parse_ranges(input_str)
  potential_ids = generate_invalid_ids(max_id)
  
  found_invalid_ids = Set.new

  potential_ids.each do |invalid_id|
    ranges.each do |start_val, end_val|
      if invalid_id >= start_val && invalid_id <= end_val
        found_invalid_ids.add(invalid_id)
        break
      end
    end
  end
  
  found_invalid_ids.sum
end

def main
  input_file_path = File.join(__dir__, 'input.txt') # Relative to this script's directory
  input_str = File.read(input_file_path)
  
  result = solve(input_str)
  puts "The sum of all invalid IDs is: #{result}"
end

main
